File: src/main/java/io/github/shaksternano/mediamanipulator/graphics/drawable/HorizontalCompositeDrawable.java
package io.github.shaksternano.mediamanipulator.graphics.drawable;

import java.awt.*;
import java.util.Objects;

public class HorizontalCompositeDrawable extends ListCompositeDrawable {

    @Override
    public void draw(Graphics2D graphics, int x, int y) {
        for (Drawable part : getParts()) {
            part.draw(graphics, x, y);
            x += part.getWidth(graphics);
        }
    }

    @Override
    public int getWidth(Graphics2D graphicsContext) {
        int totalWidth = 0;
        for (Drawable part : getParts()) {
            totalWidth += part.getWidth(graphicsContext);
        }

        return totalWidth;
    }

    @Override
    public int getHeight(Graphics2D graphicsContext) {
        int maxHeight = 0;
        for (Drawable part : getParts()) {
            maxHeight = Math.max(maxHeight, part.getHeight(graphicsContext));
        }

        return maxHeight;
    }

    @Override
    public Drawable resizeToWidth(int width) {
        throw new UnsupportedOperationException();
    }

    @Override
    public Drawable resizeToHeight(int height) {
        CompositeDrawable resized = new HorizontalCompositeDrawable();
        boolean resizedAny = false;

        for (Drawable part : getParts()) {
            try {
                part = part.resizeToHeight(height);
                resizedAny = true;
            } catch (UnsupportedOperationException ignored) {
            }

            resized.addPart(part);
        }

        if (resizedAny) {
            return resized;
        } else {
            return this;
        }
    }

    @Override
    public boolean equals(Object obj) {
        if (obj == this) {
            return true;
        } else if (obj instanceof HorizontalCompositeDrawable other) {
            return Objects.equals(getParts(), other.getParts());
        } else {
            return false;
        }
    }
}

File: core/src/main/kotlin/io/github/shaksternano/borgar/core/graphics/drawable/HorizontalCompositeDrawable.kt
package io.github.shaksternano.borgar.core.graphics.drawable

import io.github.shaksternano.borgar.core.io.closeAll
import io.github.shaksternano.borgar.core.util.kClass
import java.awt.Graphics2D
import kotlin.math.max
import kotlin.time.Duration

class HorizontalCompositeDrawable(
    private val parts: Iterable<Drawable>,
) : Drawable {

    override suspend fun draw(graphics: Graphics2D, x: Int, y: Int, timestamp: Duration) {
        var drawableX = x
        parts.forEach {
            it.draw(graphics, drawableX, y, timestamp)
            drawableX += it.getWidth(graphics)
        }
    }

    override fun getWidth(graphics: Graphics2D): Int =
        parts.fold(0) { width, part ->
            width + part.getWidth(graphics)
        }

    override fun getHeight(graphics: Graphics2D): Int =
        parts.fold(0) { height, part ->
            max(height, part.getHeight(graphics))
        }

    override fun resizeToHeight(height: Int): Drawable {
        var resizedAny = false
        val resizedParts = parts.map {
            it.resizeToHeight(height)?.also {
                resizedAny = true
            } ?: it
        }
        return if (resizedAny) {
            HorizontalCompositeDrawable(resizedParts)
        } else {
            this
        }
    }

    override suspend fun close() =
        closeAll(parts)

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (kClass != other?.kClass) return false

        other as HorizontalCompositeDrawable

        return parts == other.parts
    }

    override fun hashCode(): Int =
        parts.hashCode()

    override fun toString(): String {
        return "HorizontalCompositeDrawable(parts=$parts)"
    }
}
